### PR TITLE
Add a smaller default config option for decode

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133646

## Before
A100
| Type    |   Speedup | score_mod   | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)     |
|---------|-----------|-------------|------------|----------------|---------------------------|
| Average |     0.461 |             |            |                |                           |
| Max     |     0.996 | None        | causal     | torch.bfloat16 | (16, 16, 1, 16, 1024, 64) |
| Min     |     0.188 | None        | causal     | torch.bfloat16 | (2, 16, 1, 16, 512, 128)  |

H100
| Type    |   Speedup | score_mod   | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)     |
|---------|-----------|-------------|------------|----------------|---------------------------|
| Average |     4.528 |             |            |                |                           |
| Max     |    16.710 | None        | offset     | torch.bfloat16 | (2, 16, 1, 2, 4096, 64)   |
| Min     |     1.612 | None        | offset     | torch.bfloat16 | (16, 16, 1, 16, 512, 128) |

## After

A100:
| Type    |   Speedup | score_mod   | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)     |
|---------|-----------|-------------|------------|----------------|---------------------------|
| Average |     0.472 |             |            |                |                           |
| Max     |     1.110 | None        | causal     | torch.bfloat16 | (16, 16, 1, 16, 1024, 64) |
| Min     |     0.182 | None        | causal     | torch.bfloat16 | (2, 16, 1, 16, 4096, 128) |

H100:
| Type    |   Speedup | score_mod   | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)     |
|---------|-----------|-------------|------------|----------------|---------------------------|
| Average |     4.535 |             |            |                |                           |
| Max     |    16.691 | None        | offset     | torch.bfloat16 | (2, 16, 1, 2, 4096, 64)   |
| Min     |     1.607 | None        | offset     | torch.bfloat16 | (16, 16, 1, 16, 512, 128) |

### Failing example code

``` Python
import torch
import torch.nn as nn
import functools
from torch.nn.attention.flex_attention import flex_attention, create_block_mask


class AttentionModel(nn.Module):
    def __init__(self, initial_kv_len):
        super().__init__()
        self.kv_len = initial_kv_len
        self.q_len = 1

    def causal_mask_decode(self, b, h, q_idx, kv_idx):
        offset = self.kv_len - self.q_len
        return offset + q_idx >= kv_idx

    def forward(self, queries, keys, values, mask):
        self.kv_len = keys.shape[-2]
        bs, nh, seq_len, _ = queries.shape

        attention = functools.partial(flex_attention, block_mask=mask, enable_gqa=True)
        attention = torch.compile(attention)
        attn_output = attention(queries, keys, values)

        return attn_output


# Driver code
def main():
    # Set up parameters
    d_model = 256
    q_heads = 32
    kv_heads = 8
    kv_len = 128
    q_len = 1
    batch_size = 1

    # Initialize the model
    model = AttentionModel(kv_len)
    mask = create_block_mask(
        lambda a, b, c, d: model.causal_mask_decode(a, b, c, d), 1, 1, q_len, kv_len
    )

    # Create sample input tensors
    queries = torch.randn(batch_size, q_heads, q_len, d_model, device="cuda")
    keys = torch.randn(batch_size, kv_heads, kv_len, d_model, device="cuda")
    values = torch.randn(batch_size, kv_heads, kv_len, d_model, device="cuda")

    # Forward pass
    output = model(queries, keys, values, mask)

    print(f"Input shapes:")
    print(f"  Queries: {queries.shape}")
    print(f"  Keys: {keys.shape}")
    print(f"  Values: {values.shape}")
    print(f"Output shape: {output.shape}")


if __name__ == "__main__":
    main()

```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/kernel/flex_decoding.py [hunk 1]
```diff
@@ -302,8 +302,11 @@ def get_split_k(B: int, H: int, Mk: int, SM: int = 128) -> int:
     return split_k
 
 
-def _get_decoding_default_config(key) -> Tuple[int, int, int]:
-    default_config = (64, 2, 3)
+def _get_decoding_default_config(key) -> List[Tuple[int, int, int]]:
+    default_config = [
+        (32, 2, 3),  # Less Shared Mem
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_decoding.py:307]: keep block size at 64 and num_stages to 1 also works

**Review by Chillee — APPROVED** [review]

**Review by joydddd — APPROVED** [review]

**Comment by drisspg** [comment]
@joydddd @Chillee  The failure looks real, not sure why halving the block size hurt the numerics so much but probs something is broken

**Pull request closed without merging**